File: src/shapes/rect.rs
use image::{ImageBuffer, Rgba, RgbaImage};
use std::marker::PhantomData;
use raqote::{
    Source, SolidSource, 
    GradientStop, Point, Gradient, 
    Spread, PathBuilder, DrawTarget, 
    LineCap, LineJoin, DrawOptions, 
    StrokeStyle
};
use super::super::{DrawType, GradientSelect, Shape, Color};

pub fn draw_rect_mut(img: &mut ImageBuffer<Rgba<u8>, Vec<u8>>, shape: &Shape, _width: u32, _height: u32) {
    let (mut x, mut y) = (shape.data[0], shape.data[1]);
    let mut width = shape.data[2] as u32;
    let mut height = shape.data[3] as u32;
    if width > _width {
        width = _width;
        x = 0.;
    }
    else if width + x as u32 > _width {
        x = (_width - width) as f32;
    }

    if height > _height {
        height = _height;
        y = 0.;
    }
    else if height + y as u32 > _height {
        y = (_height - height) as f32;
    }
    // Define the start and end points for the gradient
    let source: Source;
    match &shape.color {
        Color::RGBA { red, green, blue, alpha } => {
            source = Source::Solid(SolidSource{r: *blue, g: *green, b: *red, a: *alpha});
        }
        Color::Gradients { gradients } => {
            let mut stops:Vec<GradientStop> = Vec::new();
            for gradient in gradients.iter() {
                stops.push(GradientStop {position: gradient.position, color: raqote::Color::new(gradient.color[3], gradient.color[0], gradient.color[1], gradient.color[2])});
            }
            match &gradients[0].gradient_select {
                GradientSelect::Linear {start, end} => {
                    let start = Point { x: start.x, y: start.y, _unit: PhantomData };
                    let end = Point { x: end.x, y: end.y, _unit: PhantomData };
                    source = Source::new_linear_gradient(Gradient { stops }, start, end, Spread::Pad );
                }
                GradientSelect::Radial{center, radius} => {
                    let start = Point { x: center.x, y: center.y, _unit: PhantomData };
                    source = Source::new_radial_gradient(Gradient { stops }, start, *radius, Spread::Pad );
                }
            }
        }
    }

    // Create a new path builder
    let mut path_builder = PathBuilder::new();
    path_builder.rect(0., 0., width as f32, height as f32);
        // Create a new DrawTarget
    let mut draw_target = DrawTarget::new(width as i32, height as i32);
    match shape.draw_type {
        DrawType::Fill => {
            draw_target.fill(&path_builder.finish(), &source, &DrawOptions::new());
        }
        DrawType::Stroke => {
            let style = StrokeStyle {
                width: 1.,
                cap: LineCap::Butt,
                join: LineJoin::Miter,
                miter_limit: 10.,
                dash_array: Vec::new(),
                dash_offset: 0.,
            };
            draw_target.stroke(&path_builder.finish(), &source, &style, &DrawOptions::new());
        }
    }
    // Convert raqote DrawTarget to image::ImageBuffer
    let raw_image = draw_target.get_data_u8().to_vec();
    let buffer:RgbaImage = ImageBuffer::<Rgba<u8>, Vec<u8>>::from_raw(width as u32, height as u32, raw_image).unwrap();
    for (_x, _y, pixel,) in buffer.enumerate_pixels() {
        if pixel != &Rgba([0,0,0,0]) {
            img.put_pixel(_x + x as u32, _y + y as u32, pixel.to_owned());
        }
    }
}

File: lib/shapes/rect.ts
import { Shape } from "../ffi.ts";
import { RectOptions, RectReturn } from "../types.ts";
import { setX, setY, setXAndY, setWidth, setHeight, setType, setRGB, setRGBA, setColor, setGradient } from "../utilty.ts";

export default function(data: Shape[], options?: RectOptions): RectReturn {
    if(!options) options ={};
    if(!options.x) options.x = 0;
    if(!options.y) options.y = 0;
    if(!options.color) options.color = {RGBA: {red: 0, green: 0, blue:0, alpha: 255}};
    if(!options.type) options.type = "fill";
    if(!options.width && options.width != 0) options.width = 50;
    if(!options.height && options.height != 0) options.height = 50;
    return { 
        x:options.x,
        y:options.y,
        width:options.width,
        height:options.height,
        color:options.color,
        type: options.type,
        draw(options?: RectOptions) { return draw(this, data, options) },
        setX(x) { return setX(this, x) as RectReturn },
        setY(y) { return setY(this, y) as RectReturn },
        setXAndY(x, y) { return setXAndY(this, x, y) as RectReturn },
        setWidth(width) { return setWidth(this, width) },
        setHeight(height) { return setHeight(this, height) },
        setType(type) { return setType(this, type) as RectReturn },
        setRGB(red, green, blue) { return setRGB(this, red, green, blue) as RectReturn },
        setRGBA(red, green, blue, alpha) { return setRGBA(this, red, green, blue, alpha) as RectReturn },
        setGradient(gradient) { return setGradient(this, gradient) as RectReturn },
        setColor(color) { return setColor(this, color) as RectReturn }
    }
}

function draw(rect: RectReturn, data: Shape[], options?: RectOptions): RectReturn {
    if(!options) options ={};
    if(!options.x) options.x = options.x == 0 ? 0 : options.x || rect.x;
    if(!options.y) options.y = options.y == 0 ? 0 : options.y || rect.y;
    if(!options.color) options.color = options.color ? options.color : rect.color;
    if(!options.type) options.type =  options.type ? options.type : rect.type;
    if(!options.width) options.width = options.width == 0 ? 0 : options.width || rect.width;
    if(!options.height) options.height = options.height == 0 ? 0 : options.height || rect.height;
    data.push({
        shape_type: "Rect",
        data: [options.x, options.y, options.width, options.height],
        draw_type: options.type == "fill" ? "Fill" : "Stroke",
        text: null,
        path: null,
        color: options.color
    });
    return rect
}